### DISCUSSION TITLE
Unknown symbol error due to function definition order

### DISCUSSION THREAD

**Discussion by fnaj77** [janet-lang/janet#1372] [Q&A]
Hi,
I've amassed a fairly large number of utility functions in a spork-like library that I use in personal projects.

Now that the number of functions is starting to be large, symbol definition problems are arising due to the interdependency between functions and the order in which these functions are defined in the file.

Just to be clear, this is a simple, minimal, and reproducible example of my problem:

```
(defn g []
  (f)
  (print "g"))

(defn f []
  (print "f"))

(defn main [& args]
  (g))
```

which results in:
`compile error: unknown symbol f`

I'm more than sure that a trivial problem like this has already been discussed here but, unfortunately, I can't find anything definitive about it here on github. The only thing related to my problem that I found is discussed in  #1254. From what I learned from that discussion, I could solve my problem by doing this:

```
(setdyn *redef* true)
(defn f [] '())

(defn g []
  (f)
  (print "g"))

(defn f []
  (print "f"))

(defn main [& args]
  (g))
```

which actually works, because I get the expected result but, frankly, I hope there is an automatic and idiomatic way to not have to list all the several dozen functions in my library in advance.

As always, thank you very much for your hard and amazing work on Janet!!

**Comment by iacore** [comment]
Here's how you fix the "problem".

```janet
(defn f []
  (print "f"))

(defn g []
  (f)
  (print "g"))

(defn main [& args]
  (g))
```

Janet uses early binding. The feature is very useful in finding typos at compile-time.

**Reply by fnaj77** [reply]
Hi @iacore,
thank you for taking the time for me.

I was probably unclear in explaining my problem.

Obviously I know that, in my simple example, by moving the function f before the g, I can avoid the unknown symbol error.

What I wanted to communicate with my example above is that if there are many functions, like in my case, it could be difficult to identify the precise order in which to define them to avoid the error.

**Reply by iacore** [reply]
I don't have better advice than what is in #1254.

You can reorder your functions by hand.

To do it automatically, do a topological sort.
